### PR TITLE
fix(weave_query): prevent table state loss when encountering a new logged column

### PR DESCRIPTION
**Pull request by domphan-wandb**
## Description
- Fixes [WB-17336](https://wandb.atlassian.net/browse/WB-17336)

There was a user experience issue causing a table's state (derived columns, sorting, etc) to be wiped when the user logged an additional column with their table in a subsequent run.

Initially I couldn't reproduce it, but realized the issue only occurs after the first new column is added-- subsequent added columns did not reset the table state.

**Root cause**: the **`typeShapesMatch`** util function (only used to determine if tableState should be migrated) does not handle the case of a `type` being a `union<typedDict, typedDict>` and the `toType` being a `typedDict` where the union members match the property types of the `toType`'s `typedDict`.

This fix adds a check when the `toType` is a `typedDict` and the `type` is a `union<typedDict>`. It uses the existing typedDict checking logic to confirm all the members of the union are compatible. I also refactored some of the conditional checks to make it a bit more readable.

I also need to note that the existing behavior did not correctly check the case where two unions are passed in. It will by default return true. In order to reduce the chances of shipping a regression, I've left that untouched and left a comment in the code.

https://github.com/user-attachments/assets/53384985-f791-42f1-859a-108197a9a20b

## Testing
* Unit tests
* Local dev QA

[WB-17336]: https://wandb.atlassian.net/browse/WB-17336?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: weave-js/src/components/Panel2/PanelTable/util.ts [hunk 1]
```diff
@@ -60,29 +62,60 @@ export const stripTag = (type: Type): Type => {
...
-        if (keyType === undefined || !typeShapesMatch(keyType, toKeyType)) {
-          return false;
-        }
-      }
     }
+
+    return propertyTypesAreCompatible(type, toType);
   }
+
+  // 2024/11/20, Note from Dom:
```
Comment (gtarpenning) [inline weave-js/src/components/Panel2/PanelTable/util.ts:115]: this comment could be made more actionable by describing what case we return true here

File: weave-js/src/components/Panel2/PanelTable/util.ts [hunk 2]
```diff
@@ -60,29 +62,60 @@ export const stripTag = (type: Type): Type => {
...
+    }
+    return true;
+  }
+
   type = stripTag(type);
   toType = stripTag(toType);
+
+  // List type handling
   if (isList(type) || isList(toType)) {
-    if (!isList(type) || !isList(toType)) {
```
Comment (gtarpenning) [inline weave-js/src/components/Panel2/PanelTable/util.ts:68]: weird...

File: weave-js/src/components/Panel2/PanelTable/util.ts [hunk 3]
```diff
@@ -60,29 +62,60 @@ export const stripTag = (type: Type): Type => {
...
+    // This is meant for special cases where a user adds another column to their table
+    // in a subsequent run.
+    if (isTypedDict(toType) && isUnion(type)) {
+      return type.members.every(
+        member =>
+          isTypedDict(member) &&
+          isTypedDict(toType) &&
+          propertyTypesAreCompatible(member, toType)
+      );
     }
```
Comment (gtarpenning) [inline weave-js/src/components/Panel2/PanelTable/util.ts:106]: everything else is just refactor (?), this lgtm

**Review by gtarpenning — APPROVED** [review]

File: weave-js/src/components/Panel2/PanelTable/util.ts (hunk 3 above)
Comment (domphan-wandb) [reply weave-js/src/components/Panel2/PanelTable/util.ts:106]: Yup, besides the union checking code, I refactored to make it a bit more readable. Hope it didn't make the review hard to parse. Thanks for reviewing my PRs, @gtarpenning , I appreciate it! 🙏

**Review by dannygoldstein — APPROVED** [review]
nice pr! love the thorough tests and extensive write up. this pr was a pleasure to review

my one comment is that we should rename this function from `typeShapesMatch` to something else, like `isConcattable` or something (i leave it to you to think up a better name). as it stands, this function is somewhere in between `isAssignable()` and `typeShapesMatch()`: it does not check strict type shape equality (the toType can have  less columns than the fromType), but it is also not checking assignability (it would return true for string <- number).

**Comment by domphan-wandb** [comment]
> nice pr! love the thorough tests and extensive write up. this pr was a pleasure to review
> 
> my one comment is that we should rename this function from `typeShapesMatch` to something else, like `isConcattable` or something (i leave it to you to think up a better name). as it stands, this function is somewhere in between `isAssignable()` and `typeShapesMatch()`: it does not check strict type shape equality (the toType can have less columns than the fromType), but it is also not checking assignability (it would return true for string <- number).

Thanks Danny! I'm bad at naming things, so I renamed it to `typesAreConcattable` 👍
